src/test/java/com/google/firebase/auth/internal/CryptoSignersTest.java

```diff
@@ -0,0 +1,213 @@
+      // expected
+    }
+  }
+
+  @Test
+  public void testMetadataService() throws IOException {
+    String signature = BaseEncoding.base64().encode("signed-bytes".getBytes());
+    String response = Utils.getDefaultJsonFactory().toString(
+        ImmutableMap.of("signature", signature));
+    MockHttpTransport transport = new MultiRequestMockHttpTransport(
+        ImmutableList.of(
+            new MockLowLevelHttpResponse().setContent("metadata-server@iam.gserviceaccount.com"),
+            new MockLowLevelHttpResponse().setContent(response)));
+    FirebaseOptions options = new FirebaseOptions.Builder()
+        .setCredentials(new MockGoogleCredentials("test-token"))
+        .setHttpTransport(transport)
+        .build();
+    FirebaseApp app = FirebaseApp.initializeApp(options);
+    CryptoSigner signer = CryptoSigners.getCryptoSigner(app);
+
+    assertTrue(signer instanceof CryptoSigners.IAMCryptoSigner);
+    TestResponseInterceptor interceptor = new TestResponseInterceptor();
+    ((CryptoSigners.IAMCryptoSigner) signer).setInterceptor(interceptor);
+
+    byte[] data = signer.sign("foo".getBytes());
+    assertArrayEquals("signed-bytes".getBytes(), data);
+    final String url = "https://iam.googleapis.com/v1/projects/-/serviceAccounts/"
+        + "metadata-server@iam.gserviceaccount.com:signBlob";
+    assertEquals(url, interceptor.getResponse().getRequest().getUrl().toString());
+  }
+
+  @Test
+  public void testExplicitServiceAccountEmail() throws IOException {
+    String signature = BaseEncoding.base64().encode("signed-bytes".getBytes());
+    String response = Utils.getDefaultJsonFactory().toString(
+        ImmutableMap.of("signature", signature));
+
+    // Explicit service account should get precedence
+    MockHttpTransport transport = new MultiRequestMockHttpTransport(
+        ImmutableList.of(
+            new MockLowLevelHttpResponse().setContent(response)));
+    FirebaseOptions options = new FirebaseOptions.Builder()
+        .setServiceAccount("explicit-service-account@iam.gserviceaccount.com")
+        .setCredentials(new MockGoogleCredentialsWithSigner("test-token"))
+        .setHttpTransport(transport)
+        .build();
+    FirebaseApp app = FirebaseApp.initializeApp(options);
+    CryptoSigner signer = CryptoSigners.getCryptoSigner(app);
+    assertTrue(signer instanceof CryptoSigners.IAMCryptoSigner);
+    TestResponseInterceptor interceptor = new TestResponseInterceptor();
+    ((CryptoSigners.IAMCryptoSigner) signer).setInterceptor(interceptor);
+
+    byte[] data = signer.sign("foo".getBytes());
+    assertArrayEquals("signed-bytes".getBytes(), data);
+    final String url = "https://iam.googleapis.com/v1/projects/-/serviceAccounts/"
+        + "explicit-service-account@iam.gserviceaccount.com:signBlob";
+    assertEquals(url, interceptor.getResponse().getRequest().getUrl().toString());
+
+    // Should fall back to signing-enabled credential
+    transport = new MultiRequestMockHttpTransport(
```

Comment:
Should this be split into two separate tests?

Reply:
Done